administration/src/util/qrcode.ts

```diff
@@ -1,37 +1,184 @@
+  QRCodeMatrixUtil,
+  QRCodeMode,
+  QRCodeVersion,
 } from '@zxing/library'
 import { PDFPage, rgb } from 'pdf-lib'
+import ECBlocks from '@zxing/library/esm/core/qrcode/decoder/ECBlocks'
+import MatrixUtil from '@zxing/library/esm/core/qrcode/encoder/MatrixUtil'
+import MaskUtil from '@zxing/library/esm/core/qrcode/encoder/MaskUtil'
 
-const DEFAULT_QUIET_ZONE_SIZE = 10
+const DEFAULT_QUIET_ZONE_SIZE = 4 // pt
+
+// Level 8 with EC of M gives 152 bytes
+// Level 7 with EC of L gives 154 bytes
+// Level 8 seems appropriate right now.
+export const DEFAULT_VERSION: QRCodeVersion = QRCodeVersion.getVersionForNumber(8)
+// From EC L to M we have a double of EC capability: 7% -> 15%
+export const DEFAULT_ERROR_CORRECTION: QRCodeDecoderErrorCorrectionLevel = QRCodeDecoderErrorCorrectionLevel.M
+
+function calculateBitsNeeded(
+  mode: QRCodeMode,
+  headerBitsSize: number,
+  dataBitsSize: number,
+  version: QRCodeVersion
+): number {
+  return headerBitsSize + mode.getCharacterCountBits(version) + dataBitsSize
+}
+
+function willFit(
+  mode: QRCodeMode,
+  headerBitsSize: number,
+  dataBitsSize: number,
+  version: QRCodeVersion,
+  ecLevel: QRCodeDecoderErrorCorrectionLevel
+): boolean {
+  // In the following comments, we use numbers of Version 7-H.
+  // numBytes = 196
+  const numBytes = version.getTotalCodewords()
+  // getNumECBytes = 130
+  const ecBlocks = version.getECBlocksForLevel(ecLevel)
+  const numEcBytes = ecBlocks.getTotalECCodewords()
+  // getNumDataBytes = 196 - 130 = 66
+  const numDataBytes = numBytes - numEcBytes
+  const totalInputBytes = (calculateBitsNeeded(mode, headerBitsSize, dataBitsSize, version) + 7) / 8
+  return numDataBytes >= totalInputBytes
+}
+
+function chooseMaskPattern(
+  bits: BitArray,
+  ecLevel: QRCodeDecoderErrorCorrectionLevel,
+  version: QRCodeVersion,
+  matrix: QRCodeByteMatrix
+): number {
+  let minPenalty = Number.MAX_SAFE_INTEGER // Lower penalty is better.
+  let bestMaskPattern = -1
+  // We try all mask patterns to choose the best one.
+  for (let maskPattern = 0; maskPattern < QRCode.NUM_MASK_PATTERNS; maskPattern++) {
+    MatrixUtil.buildMatrix(bits, ecLevel, version, maskPattern, matrix)
+    let penalty = calculateMaskPenalty(matrix)
+    if (penalty < minPenalty) {
+      minPenalty = penalty
+      bestMaskPattern = maskPattern
+    }
+  }
+  return bestMaskPattern
+}
+
+// The mask penalty calculation is complicated.  See Table 21 of JISX0510:2004 (p.45) for details.
+// Basically it applies four rules and summate all penalties.
+function calculateMaskPenalty(matrix: QRCodeByteMatrix): number {
+  return (
+    MaskUtil.applyMaskPenaltyRule1(matrix) +
+    MaskUtil.applyMaskPenaltyRule2(matrix) +
+    MaskUtil.applyMaskPenaltyRule3(matrix) +
+    MaskUtil.applyMaskPenaltyRule4(matrix)
+  )
+}
+
+function createHeader(mode: QRCodeMode) {
+  // This will store the header information, like mode and
+  // length, as well as "header" segments like an ECI segment.
```

Comment:
What's an ECI segment? Please add to the comment

Reply:
This code is copied from [Encoder.ts](https://github.com/zxing-js/library/blob/master/src/core/qrcode/encoder/Encoder.ts). I'll try to clarify this better by adding a comment.

I do not want to change their code to keep compatibility and avoid mistakes.